src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -449,6 +449,155 @@ ResultCode GlobalizationNative_GetSortHandle(const char* lpLocaleName, SortHandl
+                        "$Virama      = [\\p{Gujr}\\p{sc=Telu}\\p{sc=Mlym}\\p{sc=Orya}\\p{sc=Beng}\\p{sc=Deva}&\\p{Indic_Syllabic_Category=Virama}]; \n" \
+                        "$LinkingConsonant = [\\p{Gujr}\\p{sc=Telu}\\p{sc=Mlym}\\p{sc=Orya}\\p{sc=Beng}\\p{sc=Deva}&\\p{Indic_Syllabic_Category=Consonant}]; \n" \
+                        "$ExtCccZwj   = [[\\p{gcb=Extend}-\\p{ccc=0}] \\p{gcb=ZWJ}]; \n" \
+                        "$L           = [\\p{Grapheme_Cluster_Break = L}]; \n" \
+                        "$V           = [\\p{Grapheme_Cluster_Break = V}]; \n" \
+                        "$T           = [\\p{Grapheme_Cluster_Break = T}]; \n" \
+                        "$LV          = [\\p{Grapheme_Cluster_Break = LV}]; \n" \
+                        "$LVT         = [\\p{Grapheme_Cluster_Break = LVT}]; \n" \
+                        "$Extended_Pict = [:ExtPict:]; \n" \
+                        "!!chain; \n" \
+                        "!!lookAheadHardBreak; \n" \
+                        "$L ($L | $V | $LV | $LVT); \n" \
+                        "($LV | $V) ($V | $T); \n" \
+                        "($LVT | $T) $T; \n" \
+                        "[^$Control $CR $LF] ($Extend | $ZWJ); \n" \
+                        "[^$Control $CR $LF] $SpacingMark; \n" \
+                        "$Prepend [^$Control $CR $LF]; \n" \
+                        "$LinkingConsonant $ExtCccZwj* $Virama $ExtCccZwj* $LinkingConsonant; \n" \
+                        "$Extended_Pict $Extend* $ZWJ $Extended_Pict; \n" \
+                        "^$Prepend* $Regional_Indicator $Regional_Indicator / $Regional_Indicator; \n" \
+                        "^$Prepend* $Regional_Indicator $Regional_Indicator; \n" \
+                        ".;";
+
+static UChar* s_breakIteratorRules = NULL;
+static int32_t s_breakIteratorRulesLength = 0;
+
+// When doing string search operations using ICU, it is internally using a break iterator which doesn't allow breaking between some characters according to
+// the Grapheme Cluster Boundary Rules specified in http://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules.
+// Unfortunately, not all rules will have the desired behavior we need to get in .NET. For example, the rules don't allow breaking between CR '\r' and LF '\n' characters.
+// When searching for "\n" in a string like "\r\n", will get not found result.
+// We are customizing the break iterator to exclude the CRxLF rule which don't allow breaking between CR and LF.
+// The general rules syntax explained in the doc https://unicode-org.github.io/icu/userguide/boundaryanalysis/break-rules.html.
+// The ICU latest rules definition exist here https://github.com/unicode-org/icu/blob/main/icu4c/source/data/brkitr/rules/char.txt.
+static UBreakIterator* CreateCustomizedBreakIterator()
+{
+    static UChar emptyString[1];
+    UBreakIterator* breaker;
+
+    UErrorCode status = U_ZERO_ERROR;
+    if (s_breakIteratorRules != NULL)
```

Comment:
This access requires acquire semantics. Otherwise the read from *s_breakIteratorRulesLength* can be reordered to occur before the read from *s_breakIteratorRules*, which sets up a race condition:

1. Assume both statics have their default value of nullptr / 0.
2. Thread A reads *s_breakIteratorRulesLength* (value = 0).
3. Thread A is interrupted.
4. Thread B runs through this method and sets both *s_breakIteratorRulesLength* and *s_breakIteratorRules*. (The latter store uses release semantics because of the CAS operation.)
5. Thread A is resumed.
6. Thread A reads a _non-null_ value for *s_breakIteratorRules*.
7. Thread A would have asserted on line 533, but since asserts are compiled out in release builds, execution progresses to line 534 and results in incorrect behavior.

Something that may be easier is to drop the *s_breakIteratorRulesLength* static entirely and to call instead `ubrk_openRules(s_breakIteratorRules, -1, ...)`. That way this code never has to worry about accesses to these two statics being reordered since there are no longer two statics to worry about.